### PR TITLE
[omdb] Add `omdb instance info` command

### PR DESCRIPTION
**Pull request by hawkw**
Presently, `omdb` can be used to query the database for instance records
using the `omdb instances` command. However, this command does not
output all the database state associated with an instance. Instead, it
displays the instance's name, UUID, effective state, active Propolis
UUID, sled UUID, and sled serial (the last three only if the instance is
running).

This is some of the most useful information about a customer instance,
but it's not *all* of the information we have about it. Debugging issues
related to instance state management may require other information from
the instance record, as well as from the instance's active VMM record,
and its target VMM and active migration records (if it is migrating).
Adding all of these additional details to the `omdb db instances`
command is a bit of a non-starter, though. Since that command outputs a
table of all instances, with one line per instance, the output is
already *much* wider than 80 columns, and requires a pretty wide
terminal to be readable --- once the table has line wrapped, it's much
harder to interpret. Therefore, I've added an `omdb db instance info`
command that outputs a complete dump of the state of a *single*
instance. Since this command is no longer constrained to putting all
salient details about the instance on a single line in a table, we can
show every field in the instance's DB representation without having to
worry about wrapping.

Output from `omdb db instance info` looks like this (yes, I spent
entirely too long tweaking the formatting :sweat_smile:):

```console
root@oxz_switch1:~# /var/tmp/omdb-eliza-test-12 db instance info 3e9f7bcb-1ae8-4708-af4f-4b6452af3d1f
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:109::3]:32221,[fd00:1122:3344:105::3]:32221,[fd00:1122:3344:10b::3]:32221,[fd00:1122:3344:107::3]:32221,[fd00:1122:3344:108::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (98.0.0)

INSTANCE

                  ID: 3e9f7bcb-1ae8-4708-af4f-4b6452af3d1f
                name: rocky9-0
         description: application instance
          created at: 2024-08-13 16:22:18.303110 UTC
    last modified at: 2024-08-13 16:22:18.303110 UTC

CONFIGURATION

               vCPUs: 32
              memory: 96 GiB
            hostname: node0
 auto-restart policy: None

RUNTIME STATE

               state: Vmm
     last updated at: 2024-09-19T00:32:23.871767Z (generation 16)
       active VMM ID: Some(e1a87a49-7ab7-40ff-965e-4d55cc6db394)
       target VMM ID: None
        migration ID: None
        updater lock: UNLOCKED at generation: 16

   active VMM record:
     Vmm {
         id: e1a87a49-7ab7-40ff-965e-4d55cc6db394,
         time_created: 2024-09-19T00:32:23.816650Z,
         time_deleted: None,
         instance_id: 3e9f7bcb-1ae8-4708-af4f-4b6452af3d1f,
         sled_id: f15774c1-b8e5-434f-a493-ec43f96cba06,
         propolis_ip: V6(
             Ipv6Network {
                 addr: fd00:1122:3344:105::1:580,
                 prefix: 128,
             },
         ),
         propolis_port: SqlU16(
             12400,
         ),
         runtime: VmmRuntimeState {
             time_state_updated: 2024-09-19T00:32:34.197708Z,
             gen: Generation(
                 Generation(
                     4,
                 ),
             ),
             state: Running,
         },
     }
root@oxz_switch1:~#
```

I've also added `omdb db instance list` (or `instance ls`) as a subcommand of
`omdb db instance`, but these are simply aliased to the current `db instances`
command. It just seemed nice to have additional verbs.

### CODE REVIEW

**Comment by leftwo** [comment]
This is great, I've wanted instance info, but not taken the time to write it.

Should I expect it to work for destroyed instances (granted some fields will be empty)? 

```
root@oxz_switch1:~# /var/tmp/omdb-eliza-test-12 db --include-deleted --fetch-limit 999123 instance info fff916c4-814a-48a0-a04a-f887e6f81a9e
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:109::3]:32221,[fd00:1122:3344:105::3]:32221,[fd00:1122:3344:10b::3]:32221,[fd00:1122:3344:107::3]:32221,[fd00:1122:3344:108::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (98.0.0)
Error: failed to fetch details for instance fff916c4-814a-48a0-a04a-f887e6f81a9e

Caused by:
    Object (of type ById(fff916c4-814a-48a0-a04a-f887e6f81a9e)) not found: instance
```

but:
```
root@oxz_switch1:~# /var/tmp/omdb-eliza-test-12 db --include-deleted --fetch-limit 999123 instance list | grep fff916c4-814a-48a0-a04a-f887e6f81a9e
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:109::3]:32221,[fd00:1122:3344:105::3]:32221,[fd00:1122:3344:10b::3]:32221,[fd00:1122:3344:107::3]:32221,[fd00:1122:3344:108::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (98.0.0)
fff916c4-814a-48a0-a04a-f887e6f81a9e app-13                                                          destroyed <no active Propolis>                 <not on any sled>                    -    
```

File: dev-tools/omdb/src/bin/omdb/db.rs [hunk 1]
```diff
@@ -406,12 +409,38 @@ impl CliDnsGroup {
...
+
+#[derive(Debug, Subcommand)]
+enum InstanceCommands {
+    /// list instances
+    #[clap(alias = "ls")]
+    List(InstanceListArgs),
+    /// show detailed output for the selected instance.
+    ///
+    /// instances can be selected either by their UUID or by the instance and
+    /// project name.
```
Comment (leftwo) [inline dev-tools/omdb/src/bin/omdb/db.rs:425]: This comments suggests I could supply instance and project name, but I could not figure out how it parses that?

File: dev-tools/omdb/src/bin/omdb/db.rs [hunk 2]
```diff
@@ -2823,6 +2866,189 @@ async fn cmd_db_sleds(
...
+
+    // Manually format the instance struct because using its `fmt::Debug` impl
+    // will print out the user-data array one byte per line, which is horrible.
+    //
+    // /!\ WARNING /!\
+    // This does mean that anyone who adds new fields to the
+    // `nexus_db_model::Instance` type will want to make sure to update this
+    // code as well. Unfortunately, we can't just destructure the struct here to
+    // make sure this code breaks, since the `identity` field isn't public.
+    // So...just don't forget to do that, I  guess.
```
Comment (gjcolombo) [inline dev-tools/omdb/src/bin/omdb/db.rs:2904]: Should we have a similar warning over in `nexus_db_model`?

File: dev-tools/omdb/tests/usage_errors.out [hunk 3]
```diff
@@ -170,7 +171,8 @@ Commands:
   region-snapshot-replacement  Query for information about region snapshot replacements, optionally
                                manually triggering one
   sleds                        Print information about sleds
-  instances                    Print information about customer instances
+  instance                     Print information about customer instances
+  instances                    Alias to `omdb instances list`
```
Comment (gjcolombo) [inline dev-tools/omdb/tests/usage_errors.out:175]:
```suggestion
  instances                    Alias to `omdb instance list`
```

File: dev-tools/omdb/src/bin/omdb/db.rs [hunk 4]
```diff
@@ -310,8 +311,10 @@ enum DbCommands {
     RegionSnapshotReplacement(RegionSnapshotReplacementArgs),
     /// Print information about sleds
     Sleds(SledsArgs),
-    /// Print information about customer instances
-    Instances(InstancesOptions),
+    /// Print information about customer instances.
+    Instance(InstanceArgs),
+    /// Alias to `omdb instances list`.
```
Comment (gjcolombo) [inline dev-tools/omdb/src/bin/omdb/db.rs:316]:
```suggestion
    /// Alias to `omdb instance list`.
```

File: dev-tools/omdb/src/bin/omdb/db.rs [hunk 5]
```diff
@@ -2823,6 +2866,189 @@ async fn cmd_db_sleds(
...
+        STATE,
+        LAST_UPDATED,
+        LAST_MODIFIED,
+        ACTIVE_VMM,
+        TARGET_VMM,
+        MIGRATION_ID,
+        UPDATER_LOCK,
+    ]);
+    println!("\nINSTANCE\n");
+    println!("{ID:>WIDTH$}: {}", instance.id());
```
Comment (gjcolombo) [inline dev-tools/omdb/src/bin/omdb/db.rs:2942]: Any value in including the project ID here as well? (The only other missing field seems to be the cloud-init user data, but I think omitting that makes sense; there's not a lot that can be done with it from this end of the system, and if we're debugging a guest issue with the config we can always just ask the issue reporter to provide that config.)

**Review by gjcolombo — APPROVED** [review]
I like @leftwo's idea about considering displaying soft-deleted instances as such. (I might do that with an ad hoc query in OMDB, though, similar to what `cmd_db_instances` does, so as not to introduce an "ignore time_deleted" flag into the datastore.) Otherwise this LGTM with a couple small comments.

File: dev-tools/omdb/src/bin/omdb/db.rs (hunk 5 above)
Comment (hawkw) [reply dev-tools/omdb/src/bin/omdb/db.rs:2942]: Whoops, I meant to include the project ID, good catch. And, yeah, I had really wanted to just use the `fmt::Debug` representation of the `Instance` record as well, but realized that the cloud-init user data got formatted as a `Vec` of decimal numbers with each byte on its own line, which made the output more or less completely unreadable. So, all the custom formatting here is basically in service of excluding it.

**Comment by hawkw** [comment]
Regarding soft-deleted instances...the reason we currently don't report them here is because I had wanted to reuse the `instance_fetch_all` query from `nexus-db-queries`, as it's quite complex (largely because it joins with the `vmm` table twice, for the active and target VMM records). Because it's so hairy, I didn't really want to have an ad-hoc query in `omdb` that duplicated the whole thing. But, upon further consideration, it would probably be fine to use ad-hoc queries here; the reason `instance_fetch_all` is Like That is because it needs to fetch all those records in a single atomic query for consistency reasons. We don't actually have that constraint here, and it would probably be fine to just have small ad-hoc queries that fetch the instance and then separately go and get its associated VMMs and mgiration...

File: dev-tools/omdb/src/bin/omdb/db.rs (hunk 5 above)
Comment (hawkw) [reply dev-tools/omdb/src/bin/omdb/db.rs:2965]: Fixed in https://github.com/oxidecomputer/omicron/pull/6610/commits/e3013e784ef9c7039a0b3e9b141531d4794ce629. :)

**Comment by hawkw** [comment]
Okay, as of https://github.com/oxidecomputer/omicron/pull/6610/commits/e3013e784ef9c7039a0b3e9b141531d4794ce629, we now also handle deleted instances. For example:
```console
root@oxz_switch1:~# /var/tmp/omdb-eliza-test-14 db instance show fff916c4-814a-48a0-a04a-f887e6f81a9e
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:109::3]:32221,[fd00:1122:3344:105::3]:32221,[fd00:1122:3344:10b::3]:32221,[fd00:1122:3344:107::3]:32221,[fd00:1122:3344:108::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (98.0.0)

 (i) instance has been deleted

INSTANCE

                  ID: fff916c4-814a-48a0-a04a-f887e6f81a9e
          project ID: fe0da422-5c48-4b52-8010-f2fc401f090f
                name: app-13
         description: application instance
          created at: 2024-06-22 00:31:43.889412 UTC
    last modified at: 2024-06-22 00:31:43.889412 UTC
          deleted at: 2024-06-22 00:43:59.156149 UTC

CONFIGURATION

               vCPUs: 1
              memory: 2 GiB
            hostname: node13
 auto-restart policy: None

RUNTIME STATE

               state: Destroyed
     last updated at: 2024-06-22T00:43:49.440274Z (generation 5)
       active VMM ID: None
       target VMM ID: None
        migration ID: None
        updater lock: UNLOCKED at generation: 1
```

**Comment by hawkw** [comment]
Some last, unnecessary formatting tweaks:
```console
oot@oxz_switch1:~# /var/tmp/omdb-eliza-test-15 db instance show fff916c4-814a-48a0-a04a-f887e6f81a9e
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:109::3]:32221,[fd00:1122:3344:105::3]:32221,[fd00:1122:3344:10b::3]:32221,[fd00:1122:3344:107::3]:32221,[fd00:1122:3344:108::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (98.0.0)

INSTANCE
                     ID: fff916c4-814a-48a0-a04a-f887e6f81a9e
             project ID: fe0da422-5c48-4b52-8010-f2fc401f090f
                   name: app-13
            description: application instance
             created at: 2024-06-22 00:31:43.889412 UTC
       last modified at: 2024-06-22 00:31:43.889412 UTC
/!\          deleted at: 2024-06-22 00:43:59.156149 UTC

CONFIGURATION
                  vCPUs: 1
                 memory: 2 GiB
               hostname: node13
    auto-restart policy: None

RUNTIME STATE
            nexus state: Destroyed
(i)  external API state: Destroyed
        last updated at: 2024-06-22T00:43:49.440274Z (generation 5)
          active VMM ID: None
          target VMM ID: None
           migration ID: None
           updater lock: UNLOCKED at generation: 1
root@oxz_switch1:~# /var/tmp/omdb-eliza-test-15 db instance show 480c6167-dac2-4df7-808d-c9e776c1c312
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:109::3]:32221,[fd00:1122:3344:105::3]:32221,[fd00:1122:3344:10b::3]:32221,[fd00:1122:3344:107::3]:32221,[fd00:1122:3344:108::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (98.0.0)

INSTANCE
                     ID: 480c6167-dac2-4df7-808d-c9e776c1c312
             project ID: fe0da422-5c48-4b52-8010-f2fc401f090f
                   name: mysql-3
            description: MySQL Instance
             created at: 2024-09-16 00:43:10.264333 UTC
       last modified at: 2024-09-16 00:43:10.264333 UTC

CONFIGURATION
                  vCPUs: 4
                 memory: 16 GiB
               hostname: mysql-3
    auto-restart policy: None

RUNTIME STATE
            nexus state: Vmm
(i)  external API state: Stopping
        last updated at: 2024-09-19T23:15:50.698207Z (generation 5)
          active VMM ID: Some(51babee3-0ee5-4f67-bcbb-2189a39479a3)
          target VMM ID: None
           migration ID: None
           updater lock: UNLOCKED at generation: 4

      active VMM record:
          Vmm {
              id: 51babee3-0ee5-4f67-bcbb-2189a39479a3,
              time_created: 2024-09-19T23:15:50.647454Z,
              time_deleted: None,
              instance_id: 480c6167-dac2-4df7-808d-c9e776c1c312,
              sled_id: 71def415-55ad-46b4-ba88-3ca55d7fb287,
              propolis_ip: V6(
                  Ipv6Network {
                      addr: fd00:1122:3344:10b::1:4be,
                      prefix: 128,
                  },
              ),
              propolis_port: SqlU16(
                  12400,
              ),
              runtime: VmmRuntimeState {
                  time_state_updated: 2024-09-19T23:18:03.015643Z,
                  gen: Generation(
                      Generation(
                          4,
                      ),
                  ),
                  state: Stopping,
              },
          }
root@oxz_switch1:~#
```

I think this is pretty good now if either of y'all want to review it again.

**Review by leftwo — APPROVED** [review]
Thanks for showing the deleted instances too

**Pull request merged**